docs/proposals/20191017-control-plane.md

```diff
@@ -0,0 +1,494 @@
+- Scale up operations *could* be blocked based on the Health status
+  - for stacked etcd quorum must be considered
+  - for external etcd only availability of etcd should be considered.
+
+![controlplane-init-6](images/controlplane/controlplane-init-6.png)
+
+##### Scale Down
+
+- Scale down a stacked control plane
+  - For each replica being deleted:
+    - Remove the replica’s etcd member from the etcd cluster
+    - Update the kubeadm generated config map
+- Scale down a control plane with an external etcd
+  - Same as above minus Etcd management
+- Scale down operations should not be done in conjunction with an upgrade operation, this should not impact manual operations for recovery.
+- Scale down operations *could* be blocked based on the Health status
+  - for stacked etcd quorum must be considered
+  - for external etcd only availability of etcd should be considered.
+- Scale to 0 must be rejected for the initial support
+
+![controlplane-init-7](images/controlplane/controlplane-init-7.png)
+
+##### Delete of the entire ControlPlane (kubectl delete controlplane my-controlplane)
+
+- Completely removing the control plane and issuing a delete on the underlying machines.
+  - Deletion is equivalent to a scale to 1 followed by a deletion of a single replica Control Plane.
+
+##### Cluster upgrade (using create-swap-and-delete)
+
+- Triggered on any changes to Version, InfrastructureTemplate, or the KubeadmConfiguration used.
+  - Must be able to trigger an upgrade for the underlying machine image (e.g. nothing in the above list has changed but an upgrade is required)
+    - Possible implementations include (not necessary to hammer out exact implementation right now):
+      - Manually bumping a Generation field on the ControlPlane object
+      - Adding a field that is not the Generation field for users to manually bump to trigger an upgrade
+      - Adding a timestamp field, something like `nextUpgrade` which can be set to a specific time in the future (far future if no upgrades are desired) and an upgrade will run when that timestamp is passed. Good for scheduling upgrades/SLOs & quickly modifying to NOW or yesterday if an upgrade is required immediately. Would potentially update the field value to some interval (2 months from now?)
+- An upgrade will look like this:
+  - Serially go through each control plane replica not at the current config hash
+    - Check health of any replicas for the current config hash
+    - Uniquely mark the existing replica machine through setting an annotation and persisting the change (controlplane.cluster.x-k8s.io/selected-for-upgrade)
+    - Check cluster health
+    - Provision a new machine at the correct version
+    - Poll on health of replicas with the current config hash
+    - Poll on cluster health with the correct replica number
+    - Delete the marked controlplane machine
+- Upgrade of worker nodes is deferred to the MachineDeployment controller
```

Comment:
This is a UX issue that should be addressed in the Risks and mitigation section

Reply:
This is out of scope for the Control Plane management and does not represent any changes to the existing behavior.

Reply:
Let's outline what you just said either in R&M or in the non-goal / future work section

Reply:
I'd rather just remove this statement all together if it is causing confusion. There is no risk and mitigation required for something that is out of scope (worker nodes are not part of the control plane). This was previously scoped to the Upgrade CAEP that is no longer around.

Reply:
Agree that it doesn't make sense to put in risks & mitigations. +1 to removing entirely, or, at most, listing it as a non-goal.